### PR TITLE
feat(ts): implement FastPFOR codec with optimized decoder

### PR DESCRIPTION
**Pull request by DoFabien**
# feat(ts): implement FastPFOR codec with optimized decoder

## Summary

This PR implements the **SIMD-FastPFOR** integer codec in TypeScript as a reference-aligned port of the Java/C++ `CompositeCodec(FastPFOR, VariableByte)` design.

It adds both **encoding + decoding** (symmetry), plus tests to validate correctness and compatibility.

## Why

- The spec lists "SIMD-FastPFOR" as *work in progress* for JS.
- Having a native TS implementation removes the need for WASM for many consumers and helps validate the tile format end-to-end.
- A symmetric implementation (encode + decode) makes it easier to test, debug, and reason about bit ordering and stream layout.

## What's included

| File | Description |
|------|-------------|
| `ts/src/fastPforCodec.ts` | Core codec: `FastPfor`, `VariableByte`, `Composition` classes |
| `ts/src/encoding/integerEncodingUtils.ts` | `encodeFastPfor()` wired in |
| `ts/src/decoding/integerDecodingUtils.ts` | `decodeFastPfor()` wired in |
| `ts/src/decoding/fastPforCodec.reference.spec.ts` | C++ reference vector tests |
| `ts/src/encoding/fastPforEncodingUtils.spec.ts` | Round-trip encode/decode tests |
| `ts/bench/fastpfor.bench.spec.ts` | Benchmark harness |

## Correctness / Compatibility

- **Round-trip tests** (encode → decode) on multiple datasets (including multi-page streams).
- **Reference-vector compatibility tests** based on `cpp/test/test_fastpfor.cpp` (TS decoder output matches the reference vectors).

## Benchmarks

Benchmarks compare `decodeFastPfor` vs `decodeVarintInt32`:

```bash
cd ts
npm run bench:fastpfor
```

Include the larger multi-page dataset:

```bash
cd ts
npm run bench:fastpfor -- --big
```

> [!TIP]
> All benchmark datasets are **synthetic** and generated in `bench/fastpfor.bench.spec.ts`:
> - `sequential_1024`: 0, 1, 2, ... (monotonic)
> - `small_3bit_4096`: values 0–7 (low entropy)
> - `exceptions_4096`: mostly 7s with a few large outliers
> - `random_u16_8192`: pseudo-random 16-bit values (high entropy)
> - `multi_page_66000`: 66k values spanning 2 FastPFOR pages (page size = 65536)

**Results (Node v22.16.0):**

| Dataset | FastPFOR bytes | Varint bytes | FastPFOR ops/sec | Varint ops/sec |
|---------|----------------|--------------|------------------|----------------|
| sequential_1024 | 1,208 | 1,920 | 233,577 | **241,086** |
| small_3bit_4096 | 1,584 | 4,096 | **123,119** | 79,731 |
| exceptions_4096 | 1,616 | 4,105 | **123,788** | 78,658 |
| random_u16_8192 | 16,464 | 22,385 | **45,461** | 22,397 |
| multi_page_66000 (`--big`) | 102,872 | 131,104 | **4,530** | 3,079 |

Interpretation:
- **Compression:** FastPFOR is consistently smaller (22% to 61% reduction) across all datasets.
- **Decode throughput:** With the latest decoder hot-path optimizations, FastPFOR now wins on 4/5 datasets: ~54% faster on `small_3bit_4096`, ~57% faster on `exceptions_4096`, ~103% faster on `random_u16_8192`, and ~47% faster on `multi_page_66000`. It remains slightly behind on `sequential_1024` (~3% slower).
- **Conclusion:** FastPFOR gives guaranteed compression gains, and (with the current implementation) also strong decode throughput for mixed and higher-entropy datasets; Varint remains extremely efficient for simple monotonic counters.

### Optimization cost / benefit (vs no optimization)

This comparison isolates the most recent decoder hot-path optimization in `ts/src/fastPforCodec.ts` (specialized bit-unpack dispatch for common bit widths).

**Benefit (decodeFastPfor ops/sec, higher is better):**

| Dataset | Baseline | Optimized | Δ |
|---------|----------|-----------|---|
| sequential_1024 | 202,389 | 233,577 | +15% |
| small_3bit_4096 | 67,706 | 123,119 | +82% |
| exceptions_4096 | 72,216 | 123,788 | +71% |
| random_u16_8192 | 24,145 | 45,461 | +88% |
| multi_page_66000 (`--big`) | 2,833 | 4,530 | +60% |

**Cost (complexity / size):**
- `ts/src/fastPforCodec.ts` grows by **+652 LOC** (775 → 1,427) and **+30 KB** (28,385 → 58,622 bytes) — mostly auto-generated unrolled bit-unpack functions for bitwidths 1–17.

Benchmarks are intentionally kept out of `npm test` / CI.

## How to test

```bash
cd ts
npm run build
npm run lint
npm test
```

Run benchmarks:

```bash
npm run bench:fastpfor          # standard datasets
npm run bench:fastpfor -- --big  # include multi-page (66k values)
```

All 305 tests pass.

> [!NOTE]
> Round-trip testing is important for format validation — see discussion in [#740](https://github.com/maplibre/maplibre-tile-spec/pull/740).

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for taking the time to open this PR.
I think most folks will be on vacation until the first week of January. 
Regardless of the othe PR which we would need to discuss how to combine together this PR is missing a lot of coverage. 
You can wait with the tests coverage after we decide how to push forward this feature, but the end goal is to keep coverage high.

**Comment by mactrem** [comment]
Great work again!

A few observations on the benchmark results:

**Sequential performance:** The comparable performance of varint (likely due to predictable access patterns and better branch prediction) on sequential monotonic data is expected. Since such data is handled by `SequenceVector` in MLT, this case isn’t particularly relevant for our use case.

**Random data performance:** The ~2× speedup on random data is very good, especially since coordinate decoding is a dominant bottleneck in vector tile transcoding. In Java (using a realistic user session with the scalar implementation, not SIMD), I observed roughly 3× performance gains for vertex decoding. Achieving ~2× in TypeScript is excellent progress.

**Justification for optimization:** Integers are typically the most prevalent data type in tiles, so fast integer decoding is crucial. The 88% throughput improvement on 16-bit random data (most representative of coordinates, which typically dominate integer streams) strongly justifies the additional code size from explicit auto-generated, unrolled bit-unpacking.

@DoFabien Could you provide the size comparison under Gzip and Brotli compression? This would help quantify the actual bundle size impact in production.

**Comment by DoFabien** [comment]
Hi @HarelM, @mactrem,

Thank you both for taking the time to review this PR! 🙏

### Regarding Test Coverage (@HarelM)

Good news! I've added comprehensive tests to improve coverage:

| Metric | Before | After |
|--------|--------|-------|
| **Overall** | 75% | **84.66%** ✅ |
| **fastPforCodec.ts** | 84% | **96%** ✅ |

**47 new tests added:**
- `integerEncodingUtils.spec.ts`: 22 tests (varint, zigzag, RLE, delta encodings)
- `fastPforBitwidth.spec.ts`: 25 tests (all bitwidths 1-20, edge cases)

> **Note:** These tests were generated ~95% by AI, but they seem to be OK. All tests pass.

I completely understand that most folks will be on vacation until January, no rush at all! I'll be available to address feedback whenever the team is ready.

---

### Bundle Size Comparison (@mactrem)

Here are the final measurements:

| Metric | `main` | This PR | Delta |
|--------|--------|---------|-------|
| **Raw (minified)** | 42.0 KB | 57.7 KB | **+15.7 KB (+37%)** |
| **Gzip** | 11.4 KB | 15.5 KB | **+4.1 KB (+36%)** |
| **Brotli** | 10.0 KB | 13.4 KB | **+3.4 KB (+34%)** |

---

### 📝 Latest Changes (since initial PR)

Following your feedback about bundle size, I applied two optimizations:

#### What Changed

1. **Removed specialized functions for bitwidths 13, 14, 15, and 17**
   - These are rare in vector tiles
   - They now use the generic loop (acceptable performance trade-off)
   
2. **Refactored unrolled code for better compression**
   - Changed `out[outPos + N] =` → `out[op++] =` pattern
   - This creates more repetitive code that Gzip/Brotli compresses better
   - Hey, 0.4 KB saved is 0.4 KB saved! 🎉 And the `op++` pattern might also be slightly faster in JS (sequential memory writes FTW).

#### Why

The initial implementation had 17 unrolled functions (all bitwidths 1-17). This was optimal for raw performance but added significant bundle size. The "Extended Hybrid" approach keeps only the most critical bitwidths:
- **1-12**: Ultra-frequent (small integers, flags, RLE, geometry deltas)
- **16**: Critical for raw coordinates
- **13-15, 17+**: Rare, acceptable to use generic loop

#### Bundle Size Evolution

| Version | Brotli | Delta vs main |
|---------|--------|---------------|
| Initial PR (Full 1-17) | 14.5 KB | +4.5 KB |
| **This version (Hybrid 1-12+16 + op++)** | **13.4 KB** | **+3.4 KB** |
| **Savings** | | **-1.1 KB (-24%)** |


### Conclusion

I believe this is the best compromise:
- ✅ **+3.4 KB (Brotli)** bundle increase , reasonable for a tile decoding library
- ✅ **2x faster** on coordinate-heavy data (16-bit)
- ✅ **~60% faster** on typical integer streams (3-12 bit)
- ⚠️ Rare bitwidths (13-15, 17+) use generic loop , acceptable trade-off

Happy holidays to everyone! 🎄

File: ts/src/encoding/integerEncodingUtils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        const offset = new IntWrapper(0);
+
+        encodeVarintInt32Value(300, buf, offset);
+        expect(offset.get()).toBeGreaterThan(1);
+    });
+
+    it("encodes array of varint int32 values", () => {
+        const values = new Int32Array([0, 1, 127, 128, 16383, 16384]);
+        const encoded = encodeVarintInt32(values);
+        expect(encoded.length).toBeGreaterThan(0);
```
Comment (HarelM) [inline ts/src/encoding/integerEncodingUtils.spec.ts]: This test isn't really checking anything.
Please go over all the submitted tests and make sure they test something.

**Comment by HarelM** [comment]
Since a lot of this code is AI generated I suggest you review it first, do the relevant corrections and let me know when you are familiar enough with the code and can "defend" it in a code review.

**Comment by DoFabien** [comment]
My bad regarding the tests @HarelM, you're right. I used AI to generate them and didnt check closely enough, so I missed that placeholder.

Just to clarify on the unrolled source code: I used a Node.js script to generate those functions (to keep the bit logic consistant across widths).

I'll have a proper look at the test coverage later tonight or tomorrow evening. I'll replace the weak tests with actual roundtrip validation to be sure everything behaves as expected.

Will ping you when its pushed.

**Comment by DoFabien** [comment]
@HarelM Thanks for the feedback, agreed, the previous tests were too permissive.

I’ve refactored the tests into **two focused files** to ensure strict correctness without redundancy:

| File | Purpose |
|------|---------|
| `fastPforCrossLanguage.spec.ts` | Cross-language validation & integration |
| `fastPforEncodingUtils.spec.ts` | Wire format & hardening |

### Cross-language validation
- **C++ -> TS compatibility**: parses `compressed1..4` directly from `test_fastpfor.cpp` and validates bit-exact decoding.
- **Full bitwidth coverage (1–32)**: forced `maxBits` injection exercises all unrolled and generic paths.
- **Determinism**: identical encodings on consecutive calls and stable double round-trips.
- **Edge cases**: outliers/exceptions, all-zeros (b=0), constant values.
- **Anti-garbage**: all round-trips assert full buffer consumption.

### Wire format & hardening
- Validates `bigEndianBytesToInt32s` padding behavior.
- Explicit checks for the `alignedLength` header (VB-only, single block, multi-page).
- Rejects corrupted headers (negative or non-multiple-of-256 values).

### Codec hardening
Additional validation was added to `uncompress`:
- guards on header availability and validity,
- rejects negative / misaligned / out-of-bounds `outLength`.

I also added a small C++ helper (`cpp/tool/generate-fastpfor-vectors.cpp`) to regenerate reference vectors when needed (local use only, no impact on the build).  
Happy to remove it if you’d prefer to keep the repo minimal.

Ready for re-review.

**Comment by HarelM** [comment]
Please resolve the conflicts as I have just pushed a big refactoring to main.

File: cpp/tool/generate-fastpfor-vectors.cpp
Comment (HarelM) [inline cpp/tool/generate-fastpfor-vectors.cpp:1]: This looks like an overkill, I would recommend generating the files once and checking them in to the test folder for usage from different decoders.

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
+/**
+ * Cross-Language Validation Tests for FastPFOR
+ *
+ * These tests validate that the TypeScript encoder/decoder produces
+ * results compatible with the C++ reference implementation.
+ *
+ * Strategy:
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: I would recommend avoiding this strategy and instead have encoded and decoded binary references like we have for the mlt tiles, you can compare those instaed of running cpp code in the typescript tests.

**Comment by HarelM** [comment]
Please take a look at the PR I pushed in terms of how I expect to see the unit test related to this feature - more tests that are of type value - encode - decode - check equality. 
Once you have those in place please check out the coverage for the cure you added to make sure the tests cover the relevant scenarios. 
Thanks!

File: ts/src/encoding/fastPforCrossLanguage.spec.ts (hunk 2 above)
Comment (DoFabien) [reply ts/src/encoding/fastPforCrossLanguage.spec.ts]: That makes sense, agreed.
I will switch to pre-generated binary fixtures (encoded and decoded buffers) committed to the repo, similar to how MLT tile fixtures are handled, and remove the C++ parsing from the TypeScript tests.
The existing generator already produces deterministic output, so this should be a straightforward change.

I will push an update accordingly.

File: ts/src/encoding/integerEncodingUtils.ts [hunk 3]
```diff
@@ -662,6 +763,229 @@ export function encodeRleDeltaInt32(values: Int32Array | number[]): {
...
+    const numRuns = lengths.length;
+    const resultData = new BigInt64Array(numRuns * 2);
+    for (let k = 0; k < numRuns; k++) {
+        resultData[k] = BigInt(lengths[k]);
+        resultData[k + numRuns] = runValues[k];
+    }
+    return { data: resultData, numRuns };
+}
+
+export function encodeNullableZigZagRleInt64(
```
Comment (HarelM) [inline ts/src/encoding/integerEncodingUtils.ts]: Is this a result of abad merge? I believe I deleted this method.

File: ts/bench/run-fastpfor-bench.mjs
Comment (HarelM) [inline ts/bench/run-fastpfor-bench.mjs:1]: Let's use typescript, and run it with tsx if needed.

File: ts/src/decoding/integerDecodingUtils.spec.ts [hunk 4]
```diff
@@ -1,306 +1,301 @@
+
+// This file focuses on decoding-only invariants and non-canonical encodings.
```
Comment (HarelM) [inline ts/src/decoding/integerDecodingUtils.spec.ts]: Not sure I understand why was this changed... Can you elaborate?

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    const buffer = fs.readFileSync(filepath);
+    const values = new Int32Array(buffer.length / 4);
+    for (let i = 0; i < values.length; i++) {
+        values[i] = buffer.readInt32BE(i * 4);
+    }
+    return values;
+}
+
+/**
+ * Generate random values constrained to a specific bitwidth using LCG.
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: Is this really "random"?

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    return values;
+}
+
+function roundTrip(values: Int32Array): { encoded: Uint8Array; decoded: Int32Array } {
+    const encoded = encodeFastPfor(values);
+    const offset = new IntWrapper(0);
+    const decoded = decodeFastPfor(encoded, values.length, encoded.length, offset);
+    // Verify decoder consumed all bytes
+    expect(offset.get()).toBe(encoded.length);
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: It's best to keep the expect within the test and not in a utility method.

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    return { encoded, decoded };
+}
+
+describe("cross-language: C++ encoded → TS decoded", () => {
+    // Test all 4 reference vectors from binary fixtures
+    for (const idx of [1, 2, 3, 4]) {
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: Condiser using folders and read folders content instead of hard coding numbers in the test which will prevent easy addition of new files.

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        it(`generic fallback bitwidth ${bw}: TS roundtrip produces correct results`, () => {
+            const values = generateBitwidthValues(TWO_BLOCKS, bw);
+            const { decoded } = roundTrip(values);
+            expect(decoded).toEqual(values);
+        });
+    }
+});
+
+describe("cross-language: TS encoder determinism", () => {
+    it("produces identical encoding on consecutive calls", () => {
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: Given all the other test I find this one odd, consider removing it.

File: ts/src/encoding/fastPforCrossLanguage.spec.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        const values = new Int32Array(Array.from({ length: TWO_BLOCKS }, (_, i) => i % 1000));
+
+        const encoded1 = encodeFastPfor(values);
+        const encoded2 = encodeFastPfor(values);
+
+        expect(encoded1).toEqual(encoded2);
+    });
+
+    it("double roundtrip produces identical encodings", () => {
```
Comment (HarelM) [inline ts/src/encoding/fastPforCrossLanguage.spec.ts]: This file looks like "integration" tests file as it reads from files, I would recommend splitting the tests into two files, unit test and integration tests and move the inegration tests to a "higher" folder (hierarchicaly).

File: ts/src/encoding/integerEncodingUtils.ts (hunk 3 above)
Comment (DoFabien) [reply ts/src/encoding/integerEncodingUtils.ts]: Yes,
I audited the TS encoder surface and found all encodeNullable* variants were orphaned:
- not part of ts/src/index.ts public API
- not referenced anywhere in production code
- no symmetric decoder / not used by the spec
So I removed the 5 functions and their dedicated tests + dropped the now-unused BitVector import.

File: ts/src/decoding/integerDecodingUtils.spec.ts (hunk 4 above)
Comment (DoFabien) [reply ts/src/decoding/integerDecodingUtils.spec.ts]: I changed the structure/commentary to make the intent explicit: this file focuses on decoding-only invariants and non-canonical/edge-case encodings, while canonical encode->decode roundtrip coverage remains in `integerEncodingUtils.spec.ts`. The goal was to avoid mixing “decoder robustness” tests with roundtrip tests

File: ts/src/encoding/fastPforCrossLanguage.spec.ts (hunk 5 above)
Comment (DoFabien) [reply ts/src/encoding/fastPforCrossLanguage.spec.ts]: Renamed it to `generateDeterministicValues` and clarified the JSDoc. It now explicitly states that the values are deterministic pseudo-random, generated using a seeded LCG for reproducibility, not true or cryptographic randomness.

File: ts/src/encoding/fastPforCrossLanguage.spec.ts (hunk 6 above)
Comment (DoFabien) [reply ts/src/encoding/fastPforCrossLanguage.spec.ts]: the expect(offset.get()) is now in the test body, not in the helper.

**Comment by DoFabien** [comment]
Thanks @HarelM for the feedback. To address the review, this commit includes:

Orphan encoder removal
- Removed 5 orphan `encodeNullable*` functions (no decoder, no production usage)
- Removed unused `BitVector` imports
- The removed functions were leftovers from earlier iterations and not part of the public spec API

Benchmark refactoring
- Refactored the FastPFOR vs Varint benchmark to use only realistic multi-page datasets (65536 values)
- Removed `BENCH_BIG` flag logic and the dedicated runner script
- Added sanity checks for both FastPFOR and Varint decoders
- Simplified the `bench:fastpfor` npm script to run directly via Vitest

Test clarification and structure
- Improved test names and descriptions to make intent explicit
- Reorganized tests to clearly separate unit, integration, and benchmark concerns
- Moved cross-language tests to `test/integration/` to reflect their scope
- Added inline comments where test behavior was previously implicit

The benchmark can be run with:

```bash
npm run bench:fastpfor
```

File: ts/eslint.config.mjs [hunk 10]
```diff
@@ -39,13 +39,23 @@ export default tseslint.config(
...
+    languageOptions: {
+      parserOptions: {
+        project: './tsconfig.eslint.bench.json',
+        tsconfigRootDir: import.meta.dirname,
       },
     },
+  },
   {
-    ignores: ['dist', 'node_modules', 'coverage', '*.config.mjs', '*.config.js', '**/*.js']
+    ignores: ['dist', 'node_modules', 'coverage', '*.config.mjs', '*.config.js', '**/*.js', '**/*.mjs', '**/*.cjs']
```
Comment (HarelM) [inline ts/eslint.config.mjs]: This should be reverted.

File: ts/test/integration/fastPforCrossLanguage.spec.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ * Discover fixture vector pairs by scanning the fixtures directory.
+ * Returns array of vector indices (e.g., [1, 2, 3, 4]).
+ */
+function discoverFixtureVectors(): number[] {
+    const files = fs.readdirSync(FIXTURES_DIR);
+    const indices = new Set<number>();
+
+    for (const file of files) {
+        const match = file.match(/^vector(\d+)_compressed\.bin$/);
+        if (match) {
```
Comment (HarelM) [inline ts/test/integration/fastPforCrossLanguage.spec.ts]: This logic can be simplified, there's no need to check all of this, the test could simply fail with exception if the conditions do not match...

File: ts/test/integration/fastPforCrossLanguage.spec.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+const FIXTURES_DIR = path.resolve(__dirname, "../../../test/fixtures/fastpfor");
+
+/**
+ * TypedArray comparison helper to avoid relying on deep-equality behavior
+ * across different ArrayBuffer-backed types (e.g. Int32Array, BigInt64Array, Float64Array).
+ */
+function expectArrayLikeEqual<T extends ArrayLike<number> | ArrayLike<bigint>>(actual: T, expected: T) {
```
Comment (HarelM) [inline ts/test/integration/fastPforCrossLanguage.spec.ts]: I think you can use `toEqual` can't you?

File: ts/src/encoding/integerEncodingUtils.spec.ts
Comment (HarelM) [inline ts/src/encoding/integerEncodingUtils.spec.ts:1]: I think most of the tests here were moved from another file.
I suggest to open a different PR if you want to move tests around to make the review of this PR simpler.

File: ts/bench/fastpfor.bench.spec.ts
Comment (HarelM) [inline ts/bench/fastpfor.bench.spec.ts:1]: If this is not intended to run as part of the CI why keep this code? How does it help?

Comment (DoFabien) [reply ts/bench/fastpfor.bench.spec.ts:1]: It’s a developer only benchmark, not a CI test. It’s meant to manually assess performance and compression on realistic workloads when iterating on FastPFOR, without introducing non-deterministic CI results

File: ts/test/integration/fastPforCrossLanguage.spec.ts (hunk 12 above)
Comment (DoFabien) [reply ts/test/integration/fastPforCrossLanguage.spec.ts]: Yes, I used the explicit element-wise comparison to avoid relying on deep-equality semantics across different TypedArray types, but `toEqual` should be fine here. I will switch back to it.

File: ts/src/encoding/integerEncodingUtils.spec.ts
Comment (DoFabien) [reply ts/src/encoding/integerEncodingUtils.spec.ts:1]: you’re right, most of these tests were moved from another file.
I grouped them here to clarify intent and separate unit vs integration tests, based on the earlier review comments. If that makes the PR harder to review, I can move the pure test reorganization to a follow-up PR.

File: ts/src/fastPforCodec.ts [hunk 13]
```diff
@@ -0,0 +1,1335 @@
...
+        masks[bitWidth] = bitWidth === 32 ? 0xffffffff : 0xffffffff >>> (32 - bitWidth);
+    }
+    return masks;
+})();
+
+const OVERHEAD_OF_EACH_EXCEPT = 8;
+const DEFAULT_PAGE_SIZE = 65536;
+const BLOCK_SIZE = 256;
+
+interface Int32Codec {
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts]: Other parts of the code are using "regular" methods without any interface, composition etc.
They also use the terms `decode` and `encode` instead of compress and uncompress.
So I think it would be wise to keep these.
Using method instead of classes help keep the bundle size smaller in case someone only needs either the decoding or the encoding.

File: ts/src/encoding/integerEncodingUtils.spec.ts
Comment (HarelM) [reply ts/src/encoding/integerEncodingUtils.spec.ts:1]: Yes, let's make this PR focused on FastPFor and leave ordering of tests to a different PR(s).

File: ts/bench/fastpfor.bench.spec.ts
Comment (HarelM) [reply ts/bench/fastpfor.bench.spec.ts:1]: Since I don't see who will run it in the future I advise to remove it.
Once the implementation is complete and the results are OK from a performance perspective I don't see a reason to keep this file.
It will only make maintenance harder with minimal value at best.
We can write this code inside an issue somewhere to keep it for future uses, but avoid adding it to source control to reduce maintenance...

**Comment by DoFabien** [comment]
Done. I’ve taken your review comments into account and implemented the requested changes to align the FastPFOR implementation with the existing style and constraints.

I also have a more aggressively optimized version (inlined bit-unpacking via switch cases) showing up to ~3× speedup vs varint in some cases. Would you prefer this as a follow-up PR, or should I include it here?

File: ts/src/fastPforCodec.ts [hunk 14]
```diff
@@ -0,0 +1,1310 @@
...
+
+    const byteContainerSize = (3 * pSize) / BLOCK_SIZE + pSize;
+    let byteContainer = new Uint8Array(byteContainerSize) as Uint8Buf;
+
+    const initialPackedSize = (pSize / 32) * 4;
+    for (let k = 1; k < dataToBePacked.length; k++) {
+        dataToBePacked[k] = new Int32Array(initialPackedSize);
+    }
+
+    function getBestBFromData(inValues: Int32Array, pos: number): void {
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:740]: Can this method be extracted outside the `createFastPforEncoder`?

Comment (HarelM) [reply ts/src/fastPforCodec.ts:740]: And the other methods as well?

File: ts/src/fastPforCodec.ts [hunk 15]
```diff
@@ -0,0 +1,1310 @@
...
+            const high = currentWord & highMask;
+
+            const value = (low | (high << lowBits)) & mask;
+            out[outPos + i] = value | 0;
+            bitOffset = bitWidth - lowBits;
+        }
+    }
+}
+
+function createFastPforEncoder(pageSize = DEFAULT_PAGE_SIZE) {
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:724]: Why do we need this method? Does the encoder has a state? I would image it should be a pure function.
Consider exporting only the encode method.

Also I would advise, if possible to split the encode and decode to separate files, this is what we did for other encode decode.

File: ts/src/fastPforCodec.ts [hunk 16]
```diff
@@ -0,0 +1,1310 @@
...
+
+    return out.subarray(0, outPos.get());
+}
+
+export function decodeFastPforInt32(encoded: Int32Buf, numValues: number): Int32Array {
+    const inPos = new IntWrapper(0);
+    const outPos = new IntWrapper(0);
+    const decoded = new Int32Array(numValues);
+
+    // 1. FastPFOR uncompress
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:1248]: "decode" instead of "uncompress" or consider removing the comment completely.

File: ts/src/fastPforCodec.ts [hunk 17]
```diff
@@ -0,0 +1,1310 @@
...
+                    `FastPFOR: outLength=${outLength} exceeds output capacity ${out.length - outPos.get()}`,
+                );
+            }
+
+            headlessUncompress(inValues, inPos, inLength, out, outPos, outLength);
+        },
+    };
+}
+
+function encodeVarint(
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:1143]: I think this method exists elsewhere, doesn't it?

File: ts/src/fastPforCodec.ts [hunk 18]
```diff
@@ -0,0 +1,1310 @@
...
+        const v = bytes[i] | (bytes[i + 1] << 8) | (bytes[i + 2] << 16) | (bytes[i + 3] << 24) | 0;
+        out[outIdx++] = v;
+    }
+
+    outPos.set(outIdx);
+    inPos.add(inLength);
+    return out;
+}
+
+function decodeVarint(
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:1180]: Same as above, this looks familiar...

File: ts/src/fastPforCodec.ts [hunk 19]
```diff
@@ -0,0 +1,1310 @@
...
+                    }
+                }
+            }
+        }
+
+        outPos.set(tmpOutPos);
+        inPos.set(inExcept);
+    }
+
+    function headlessUncompress(
```
Comment (HarelM) [inline ts/src/fastPforCodec.ts:1093]: This still uses the terminology of uncompress instead of decode.

**Comment by HarelM** [comment]
I've added a few more comments, please try and look at the code throughly and fix other issues you see, not just those I mentioned...

**Comment by DoFabien** [comment]
Got it. To be honest, the current version is not yet satisfactory. 
I am doing a full line-by-line refactor of the entire implementation to address this. I am adding all necessary edge cases and hardening the logic to make it much more robust.

My goal is to ensure the code is better organized, better structured, and thoroughly documented, so it is easy to review and maintain. I am also integrating several optimizations that will further improve performance.
This process takes time to get every detail right, so I will push the updated commit in a few days once I am fully satisfied that it meets the required quality standards.

**Comment by HarelM** [comment]
I would recommend considering splitting those changes to as small as PR is possible but still makes sense for easy review. 
The bigger it is the harder it is to review and follow up on comments.

**Comment by DoFabien** [comment]
Closing this PR in favor of a cleaned-up rewrite split into two stacked PRs:

- #762 — core FastPFOR int32 codec implementation (TypeScript decoder/unpack/spec + unit tests)
- #763 — binary fixtures + cross-language integration tests (C++ vectors -> TS decode)

Together, #762 + #763 replace #758 with a more focused, more readable and more robust implementation (same wire format)

Please continue the review/discussion on #762 and #763. 
No further updates will be pushed to #758.
